### PR TITLE
On exit, the driver will automaticallly shutdown any nodes which were…

### PR DESCRIPTION
**Pull request by shamsasari**
…n't waited for.

The motivation for this came with the recent change that a default notary is started by the driver, which if ignored will leak the notary process.

This change removes the need for waitForAllNodesToFinish().

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 1]
```diff
@@ -724,12 +724,15 @@ abstract class AbstractNode(val configuration: NodeConfiguration,
...
-            NetworkMapCacheImpl(
+            val cache = NetworkMapCacheImpl(
                     PersistentNetworkMapCache(
                             this@AbstractNode.database,
                             this@AbstractNode.configuration,
                             networkParameters.notaries),
-                    identityService)
+                    identityService
+            )
+            runOnStop += cache::close
```
Comment (exFalso) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: perhaps put this into a `makeNetworkMapCache()` mirroring the rest of these initialisations

File: samples/irs-demo/cordapp/src/test/kotlin/net/corda/irs/Main.kt [hunk 2]
```diff
@@ -20,7 +20,5 @@ fun main(args: Array<String>) {
         startWebserver(controller)
         startWebserver(nodeA)
         startWebserver(nodeB)
-
-        waitForAllNodesToFinish()
```
Comment (exFalso) [inline samples/irs-demo/cordapp/src/test/kotlin/net/corda/irs/Main.kt]: If we remove this won't the driver immediately shut down?

Comment (shamsasari) [reply samples/irs-demo/cordapp/src/test/kotlin/net/corda/irs/Main.kt]: Doh, yes of course! I'll add it back, but then the method is at odds with this new feature. And also it's only meant to be called at the end of the dsl block? I wonder if it even makes sense to use the driver in these scenarios and instead use the new cordformation thing.

Comment (exFalso) [reply samples/irs-demo/cordapp/src/test/kotlin/net/corda/irs/Main.kt]: Agreed. Hmm then these Main.kts should be removed. Perhaps ask devrel first whether these are used at all?

Comment (shamsasari) [reply samples/irs-demo/cordapp/src/test/kotlin/net/corda/irs/Main.kt]: I added a `waitForNodesToFinish` parameter instead.

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 3]
```diff
@@ -396,6 +395,7 @@ data class DriverParameters(
     fun setUseTestClock(useTestClock: Boolean) = copy(useTestClock = useTestClock)
     fun setInitialiseSerialization(initialiseSerialization: Boolean) = copy(initialiseSerialization = initialiseSerialization)
     fun setStartNodesInProcess(startNodesInProcess: Boolean) = copy(startNodesInProcess = startNodesInProcess)
+    fun setTerminateNodesOnShutdown(terminateNodesOnShutdown: Boolean) = copy(waitForNodesToFinish = terminateNodesOnShutdown)
```
Comment (exFalso) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]: Perhaps name this the same as the parameter for consistency

File: testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt [hunk 4]
```diff
@@ -585,8 +588,9 @@ class DriverDSL(
     override val notaryHandles: List<NotaryHandle> get() = _notaries
     private lateinit var networkParameters: NetworkParametersCopier
 
+
```
Comment (exFalso) [inline testing/node-driver/src/main/kotlin/net/corda/testing/driver/Driver.kt]: whitespace 🚓

**Comment by shamsasari** [comment]
Comments addressed.

**Pull request merged**